edk2toolext/environment/reporttypes/coverage_report.py

```diff
@@ -0,0 +1,337 @@
+
+            # Merge the file results
+            else:
+                to_update = file_dict[path]
+
+                # Merge lines first
+                for line in file.find("lines").iter():
+                    line_number = line.attrib.get("number")
+                    if not line_number:
+                        continue
+                    match = to_update.find("lines").find(f".//line[@number='{line_number}']")
+                    match.set("hits", str(int(match.attrib.get("hits")) + int(line.attrib.get("hits"))))
+        return file_dict
+
+    def build_inf_source_dictionary(self, db: Edk2DB, env_id: int, query: str) -> dict:
+        """Builds a dictionary of INFs and the source files they use.
+
+        Args:
+            db (Edk2DB): The database containing the necessary data
+            env_id (int): The environment id to query
+            query (str): The query to use to get the data
+
+        Returns:
+            dict[str, list[str]]: A dictionary of INFs and the source files they use.
+        """
+        entry_dict = {}
+        results = db.connection.execute(query, (env_id,)).fetchall()
+
+        for inf, source in results:
+            if inf not in entry_dict:
+                entry_dict[inf] = [source]
+            else:
+                entry_dict[inf].append(source)
+        return entry_dict
+
+    def build_report(self, db: Edk2DB, env_id: int, source_coverage_dict: dict, inf_source_dict: dict):
+        """Builds the report.
+
+        For each source file in each INF in the inf_source dictionary, look to see if there is coverage data for it in
+        the source_coverage dictionary. If it exists, insert it into the new report. Writes the final report to the
+        specified file.
+        """
+        pp_list, = db.connection.execute(PACKAGE_PATH_QUERY, (env_id,)).fetchone()
+        pp_list = pp_list.split(os.pathsep)
+
+        root = ET.Element("coverage")
+        sources = ET.SubElement(root, "sources")
+        # Set the sources so that reports can find the right paths.
+        source = ET.SubElement(sources, "source")
+        source.text = self.args.workspace
+        for pp in pp_list:
+            source = ET.SubElement(sources, "source")
+            source.text = str(Path(self.args.workspace, pp))
+
+        packages = ET.SubElement(root, "packages")
+        for path, source_list in inf_source_dict.items():
+            if not source_list:
+                continue
+            inf = ET.SubElement(packages, "package", path=path, name=Path(path).name)
+            classes = ET.SubElement(inf, "classes")
+
+            for source in source_list:
+                # Check if the file should be excluded
+                exclude_file = False
+                for pattern in self.args.exclude:
+                    if fnmatch.fnmatch(source, pattern):
+                        exclude_file = True
+                        break
+                if exclude_file:
```

Comment:
should log the exclusion like you do below?

Reply:
Added logging. Thanks!